File: src/main/java/com/zalphion/featurecontrol/FeatureControl.java
package com.zalphion.featurecontrol;

import com.zalphion.featurecontrol.bundle.FeatureBundle;
import com.zalphion.featurecontrol.bundle.FeatureBundleSerializer;
import com.zalphion.featurecontrol.lib.http.*;
import com.zalphion.featurecontrol.lib.http.cache.CacheAwareHttpFunction;
import com.zalphion.featurecontrol.lib.result.Failure;
import com.zalphion.featurecontrol.lib.result.Result;
import com.zalphion.featurecontrol.source.FeatureSource;
import lombok.val;
import org.jspecify.annotations.NonNull;

import java.io.IOException;
import java.net.URI;
import java.nio.charset.StandardCharsets;
import java.util.Collections;
import java.util.Optional;

public class FeatureControl {
    private final @NonNull URI baseUri;
    private final @NonNull HttpFunction http;

    public FeatureControl(@NonNull @lombok.NonNull URI baseUri) {
        this(baseUri, HttpUrlConnectionHttpFunction.builder().build());
    }

    public FeatureControl(@NonNull @lombok.NonNull URI baseUri, @NonNull @lombok.NonNull HttpFunction http) {
        this.baseUri = baseUri;

        // enforce cache awareness
        if (http instanceof CacheAwareHttpFunction) {
            this.http = http;
        } else {
            this.http = new CacheAwareHttpFunction(http);
        }
    }

    public static @lombok.NonNull FeatureControl northAmerica() {
        return new FeatureControl(URI.create("https://na.featurecontrol.app"));
    }

    public static @lombok.NonNull FeatureControl europe() {
        return new FeatureControl(URI.create("https://eu.featurecontrol.app"));
    }

    public static @lombok.NonNull FeatureControl asiaPacific() {
        return new FeatureControl(URI.create("https://ap.featurecontrol.app"));
    }

    public @NonNull Result<FeatureBundle> getBundle(@NonNull @lombok.NonNull String sdkKey) {
        try {
            val request = HttpRequest.builder()
                    .method(HttpMethod.GET)
                    .url(baseUri.resolve("/sdkapi/v1/bundle").toURL())
                    .header("Authorization", Collections.singletonList("Bearer " + sdkKey))
                    .build();

            val response = http.exchange(request);

            switch(response.getStatusCode()) {
                case 200:
                    val json = new String(response.getBody(), StandardCharsets.UTF_8);
                    return FeatureBundleSerializer.fromJson(json);
                case 401: return new Failure<>("Invalid SDK Key");
                case 403: return new Failure<>("Application has been banned for abuse");
                case 429: return new Failure<>("Rate limit temporarily exceeded");
                default: return new Failure<>("Unexpected status code " + response.getStatusCode() + ": " + response.getStatusReason());
            }
        } catch (IOException e) {
            return new Failure<>(Optional.ofNullable(e.getMessage()).orElse("Unknown error"));
        }
    }

    public @NonNull FeatureSource toFeatureSource(@NonNull @lombok.NonNull String sdkKey) {
        return new FeatureSource() {
            @Override
            protected @NonNull @lombok.NonNull Result<FeatureBundle> getInternal() {
                return getBundle(sdkKey);
            }
        };
    }
}

File: src/main/kotlin/com/zalphion/featurecontrol/client/FeatureControl.kt
package com.zalphion.featurecontrol.client

import com.zalphion.featurecontrol.FeatureBundle
import dev.forkhandles.result4k.Failure
import dev.forkhandles.result4k.Result4k
import dev.forkhandles.result4k.asFailure
import dev.forkhandles.result4k.asSuccess
import kotlinx.serialization.ExperimentalSerializationApi
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.decodeFromStream
import okhttp3.Cache
import okhttp3.OkHttpClient
import okhttp3.Request
import java.io.File
import java.net.URI

/**
 * Will respect all cache headers, including eTag and max-age
 */
class FeatureControl @JvmOverloads constructor(
    private val baseUri: URI,
    cacheDir: File = File.createTempFile("feature-control", "cache").apply { delete(); mkdir() }
) {
    companion object {
        @JvmStatic val northAmerica = FeatureControl(URI("https://na.featurecontrol.app"))
        @JvmStatic val europe = FeatureControl(URI("https://eu.featurecontrol.app"))
        @JvmStatic val asiaPacific = FeatureControl(URI("https://ap.featurecontrol.app"))
    }

    private val client: OkHttpClient = OkHttpClient.Builder()
        .followRedirects(false)
        .cache(Cache(cacheDir, maxSize = 10L * 1024 * 1024)) // 10 MB cache
        .build()

    @OptIn(ExperimentalSerializationApi::class)
    internal fun getBundle(sdkKey: String): Result4k<FeatureBundle, String> {
        val request = Request.Builder()
            .url(baseUri.resolve("/sdkapi/v1/bundle").toURL())
            .header("Authorization", "Bearer $sdkKey")
            .build()

        return try {
            client.newCall(request).execute().use { response ->
                when (response.code) {
                    200 -> response.body.byteStream().use {
                        Json.decodeFromStream(FeatureBundle.serializer(), it)
                    }.asSuccess()
                    401 -> "Invalid SDK Key".asFailure()
                    403 -> "Your key has been banned for abuse.  Please review the fair-use documentation".asFailure()
                    429 -> "Rate limit temporarily exceeded.  Please review the fair-use documentation".asFailure()
                    else -> "Unexpected status code: ${response.code}".asFailure()
                }
            }
        } catch (e: Exception) {
            Failure(e.message ?: "Unknown error")
        }
    }
}